src/blueprint.zig

```diff
@@ -372,6 +374,246 @@ pub const Pattern = struct {
+	var count = 0;
+
+	for(std.meta.declarations(Block)) |decl| {
+		const declInfo = @typeInfo(@TypeOf(@field(Block, decl.name)));
+		if(declInfo != .@"fn") continue;
+		if(declInfo.@"fn".return_type != bool) continue;
+		if(declInfo.@"fn".params.len > 1) continue;
+
+		tempFields[count] = .{.name = decl.name, .value = count};
+		count += 1;
+	}
+
+	const outFields: [count]std.builtin.Type.EnumField = tempFields[0..count].*;
+
+	return @Type(.{.@"enum" = .{
+		.tag_type = u8,
+		.fields = &outFields,
+		.decls = &.{},
+		.is_exhaustive = true,
+	}});
+}
+
+pub const Mask = struct {
+	const AndList = ListUnmanaged(Entry);
+	const OrList = ListUnmanaged(AndList);
+
+	entries: OrList,
+
+	const or_ = '|';
+	const and_ = '&';
+	const inverse = '!';
+	const tag = '$';
+	const property = '@';
+
+	const Entry = struct {
+		inner: Inner,
+		isInverse: bool,
+
+		const Inner = union(enum) {
+			block: Block,
+			blockType: u16,
+			blockTag: Tag,
+			blockProperty: Property,
+
+			const Property = generatePropertyEnum();
+
+			fn initFromString(specifier: []const u8) !Inner {
+				switch(specifier[0]) {
+					tag => {
+						const blockTag = specifier[1..];
+						if(blockTag.len == 0) return error.TagNotFound;
+						return .{.blockTag = Tag.get(blockTag) orelse return error.TagNotFound};
+					},
+					property => {
+						const propertyName = specifier[1..];
+						const propertyValue = std.meta.stringToEnum(Property, propertyName) orelse return error.PropertyNotFound;
+						return .{.blockProperty = propertyValue};
+					},
+					else => {
+						switch(try parseBlockLike(specifier)) {
+							.block => |block| return .{.block = block},
+							.blockType => |blockType| return .{.blockType = blockType},
+						}
+					},
+				}
+			}
+
+			fn match(self: Inner, block: Block) bool {
+				return switch(self) {
+					.block => |desired| block.typ == desired.typ and block.data == desired.data,
+					.blockType => |desired| block.typ == desired,
+					.blockTag => |desired| block.hasTag(desired),
+					.blockProperty => |blockProperty| switch(blockProperty) {
+						inline else => |prop| @field(Block, @tagName(prop))(block),
+					},
+				};
+			}
+		};
+
+		fn initFromString(specifier: []const u8) !Entry {
+			const isInverse = specifier[0] == '!';
+			const entry = try Inner.initFromString(specifier[if(isInverse) 1 else 0..]);
+			return .{.inner = entry, .isInverse = isInverse};
+		}
+
+		pub fn match(self: Entry, block: Block) bool {
+			const isMatch = self.inner.match(block);
+			if(self.isInverse) {
+				return !isMatch;
+			}
+			return isMatch;
+		}
+	};
+
+	pub fn initFromString(allocator: NeverFailingAllocator, source: []const u8) !@This() {
+		var orStorage: OrList = .{};
+		errdefer orStorage.deinit(allocator);
+
+		var oredExpressions = std.mem.splitScalar(u8, source, or_);
+		while(oredExpressions.next()) |subExpression| {
+			if(subExpression.len == 0) continue;
```

Comment:
Should this be an error? Otherwise `cubyz:air||cubyz:void` is a valid mask. Or is this intentional to allow `||` as used in many programming languages?

Reply:
This is not intentional, but such deduplication of operators should not have any performance or syntactic implications.

Reply:
Changed to error